### ISSUE TITLE
Support python 3

### ISSUE THREAD

**Issue by quicklizard99** [NaturalHistoryMuseum/inselect#95]
We are tied to Python 2.7.x because one of our dependencies - [OpenCV](http://opencv.org) - provides only Python 2.x bindings. Although Python 2.7 will be [supported until 2020](http://legacy.python.org/dev/peps/pep-0373/), there will be only bugfix releases - no new development.

Python 3 [is the future](http://legacy.python.org/dev/peps/pep-0404/). OpenCV 3 (currently in alpha) will provide [Python 3 bindings](http://opencv.org/opencv-3-0-alpha.html).

**Comment by quicklizard99** [comment]
Python 3 has a different API for C extension modules. Support for Python 3 requires that we (or someone else) rewrites the `pydmtx` module, which provides an interface to the `libdmtx` Data Matrix barcode reading library. Also the `zbar` module, if we decide to support the `zbar` barcode reading library.

**Comment by stefanv** [comment]
Both of these projects are quite old, so I guess we shouldn't hold out
too much hope that they'll port to 3.  Since `libdtmx` is written in C,
we should be able to construct a Cython wrapper for the functionality
used in inselect fairly easily.  If you think that would be useful, I'd
be happy to put some time into it.

**Comment by quicklizard99** [comment]
Thanks @stefanv - that would be great. Probably best to wait until I have tried Inselect with Python 3 - I think it likely that there will be other sticking points - `cx_freeze`, `PyInstaller`, `PySide` - to sort out before we get to barcodes. Ideally, barcode reading would be part of OpenCV because neither `libdmtx` nor `zbar` haven't been developed for years.

**Comment by stefanv** [comment]
Alright, let me know when you are done with that phase, then I 
will dive into this in more detail!

**Comment by quicklizard99** [comment]
A first pass at python 3 support on [iss95](/NaturalHistoryMuseum/inselect/tree/iss95).

**Comment by quicklizard99** [comment]
How I  built OpenCV 3.0.0 RC1 on Ubuntu 14.04 (with help from [1][1] and [2][2]). Download `opencv-3.0.0-rc1.zip` to ~/build/. The following will make and install to `~/local/opencv-3.0.0-rc1`. The `cv2` extension python will be installed to the virtual env's site-packages: `~/Envs/dig343/lib/python3.4/site-packages` for me.

```
cd ~/build
unzip opencv-3.0.0-rc1.zip
cd opencv-3.0.0-rc1
mkdir build
cd build
workon dig343    # the name of my virtual env
rm -rf * && cmake -D CMAKE_BUILD_TYPE=RELEASE \
        -D PYTHON3_EXECUTABLE=$(which python3) \
        -D PYTHON3_PACKAGES_PATH=`python3 -c 'from distutils.sysconfig import get_python_lib; print(get_python_lib())'` \
        -D PYTHON3_INCLUDE_DIR=`python3 -c 'from distutils.sysconfig import get_python_inc; print(get_python_inc())'` \
        -D PYTHON3_LIBRARY=`python3 -c 'from distutils import sysconfig; print("{LIBDIR}/{LDLIBRARY}".format(**sysconfig.get_config_vars()))'` \
        -D CMAKE_INSTALL_PREFIX=~/local/opencv-3.0.0-rc1 \
        -D BUILD_EXAMPLES=ON \
        -D INSTALL_PYTHON_EXAMPLES=ON \
        -D WITH_QT=ON \
        ~/build/opencv-3.0.0-rc1/

    make -j4 # 4 cores - use the appropriate number here
    make install
```

Test

```
python -c "import cv2; print(cv2.__version__)"
```

[1]: http://stackoverflow.com/a/24115039/1773758
[2]: http://luigolas.com/blog/2014/09/15/install-opencv3-with-python-3-mac-osx/

**Comment by stefanv** [comment]
Another few potential options for wrapping:

https://github.com/dlbeer/quirc
https://github.com/glassechidna/zxing-cpp

What kind of barcodes do we need to support?

**Comment by quicklizard99** [comment]
#372 and #374 add Python 3 compatible wrappers for `libdmtx` and `zbar` respectively.

**Comment by quicklizard99** [comment]
Things blocking Python 3
* `PySide` - not updated for years, stuck on Python 2 (and `Qt 4`) - fixed by #389
* `zbar` - not updated for years, stuck on Python 2 - fixed by #374 
* `libdmtx` - not updated for years, stuck on Python 2 - fixed by #372 
* `OpenCV 2.4` - stuck on Python 2 - Inselect needs to altered to use `OpenCV 3.1`

`PyInstaller` and `cx_Freeze` (used to freeze Inselect for Mac OS X and Windows respectively) both claim to support Python 3.

**Comment by quicklizard99** [comment]
Pull #403: all tests pass on Mac OS X, Windows and on Ubuntu 12.04, 14.04 and 16.04.

Neither the Mac OS X nor the Windows installers build properly - the next job.
